src/Microsoft.AspNet.WebUtilities/HttpRequestStreamReader.cs

```diff
@@ -0,0 +1,437 @@
+            }
+
+            if (!stream.CanRead)
+            {
+                throw new ArgumentException(Resources.HttpRequestStreamReader_StreamNotReadable, nameof(stream));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            _stream = stream;
+            _encoding = encoding;
+            _decoder = encoding.GetDecoder();
+
+            if (bufferSize < MinBufferSize)
+            {
+                bufferSize = MinBufferSize;
+            }
+
+            _byteBufferSize = bufferSize;
+            _byteBuffer = new byte[bufferSize];
+            var maxCharsPerBuffer = encoding.GetMaxCharCount(bufferSize);
+            _charBuffer = new char[maxCharsPerBuffer];
+        }
+
+        public HttpRequestStreamReader(
+            Stream stream,
+            Encoding encoding,
+            int bufferSize,
+            ArrayPool<byte> bytePool,
+            ArrayPool<char> charPool)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (!stream.CanRead)
+            {
+                throw new ArgumentException(Resources.HttpRequestStreamReader_StreamNotReadable, nameof(stream));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            if (bytePool == null)
+            {
+                throw new ArgumentNullException(nameof(bytePool));
+            }
+
+            if (charPool == null)
+            {
+                throw new ArgumentNullException(nameof(charPool));
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize));
+            }
+
+            _stream = stream;
+            _encoding = encoding;
+            _byteBufferSize = bufferSize;
+            _bytePool = bytePool;
+            _charPool = charPool;
+
+            _decoder = encoding.GetDecoder();
+
+            _byteBuffer = _bytePool.Rent(bufferSize);
+
+            try
+            {
+                var requiredLength = encoding.GetMaxCharCount(bufferSize);
+                _charBuffer = _charPool.Rent(requiredLength);
+            }
+            catch
+            {
+                _bytePool.Return(_byteBuffer);
+                _byteBuffer = null;
+
+                if (_charBuffer != null)
+                {
+                    _charPool.Return(_charBuffer);
+                    _charBuffer = null;
+                }
+            }
+        }
+
+#if dnx451 
+        public override void Close()
+        {
+            Dispose(true);
+        }
+#endif
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && _stream != null)
+            {
+                _stream = null;
```

Comment:
I assume you're just moving this code from somewhere else, but I do have a few comments on the code itself...

Dispose shouldn't null out fields, it just causes null refs in reality. Use a disposed bool.